### PR TITLE
Simplify function arg2str by mlutiple dispatch

### PR DESCRIPTION
**Pull request by ZMAlt**
The original function has too many control flows. It's hard to read. I changed it to several methods.

### CODE REVIEW

**Comment by joa-quim** [comment]
Thanks for looking in improvements, but a very quick look, please do not remove the ``::String`` annotations. I know they don't look nice but they are there to restrict the ``Any`` propagation. It's very annoying that Julia still return a ``Any`` type when doing ``string(text)`` and text is a string of type ``Any``. This all originates from the fact that the ``d`` Dictionary is of type ``{Symbol, Any}``.

**Comment by ZMAlt** [comment]
I understand your concerns. I tried my best to ensure the type of function `arg2str`'s return value is String or subtypes of AbstractString. If all options' parsing call arg2str, we may be able to terminate the `Any` type here.

Even though all the tests in `test_common_opts.jl` passed, I'm still worried about whether the tests cover all cases.

**Comment by joa-quim** [comment]
But the annotations are there to prevent ``Any`` contamination in the intermediate steps.*e.g.* ``cmd *= string(val)`` would convert a ``cmd`` of string type into a ``Any``. That's why I want to keep those type annotations. The problem, again, is that ``string(val)`` is not guarantied to return a String type.

**Comment by joa-quim** [comment]
And its weird why the test are failing to delete the ``png.png`` file. It's certainly a silly name likely resulting from some mistake or tiredness, but it's a valid name and I do see it get created  when I run the tests in my computer.

Yes, the tests may not test all ``arg2str`` cases. I also use the documentation build and the, on works, GMT classic illustrations, to test test things further. But that is not a systematic testing procedure.

**Comment by ZMAlt** [comment]
> The problem, again, is that `string(val)` is not guarantied to return a String type.

That's weird. Can you show me a case.

> Note, this all resulted when trying to reduce the number of dynamic dispatches reported by JET.

Never used Jet.jl. I'll try it later.

---

> And its weird why the test are failing to delete the png.png file. It's certainly a silly name likely resulting from some mistake or tiredness, but it's a valid name and I do see it get created when I run the tests in my computer.

CI says:
```
    SUBPLOT1
Error:  [ERROR]: Number of RowsxCols is required!
```
No png.png was generated due to the wrong parsing of subplot's `RowsxCols`. Goona fix it.

**Comment by joa-quim** [comment]
> That's weird. Can you show me a case?

We must use JET to see it, but be warned that JET is scaring. It *accuses* of hundreds of *errors* that I find most of times totally incomprehensible and happen inside Julia itself.

**Comment by joa-quim** [comment]
You still need to restore the ``::String`` type annotations.

**Comment by ZMAlt** [comment]
Let's put aside the issue of type annotations. From CI, the current changes may lead to a crash, but this doesn't always happen.

**Comment by joa-quim** [comment]
That crash is an old _annoyance_. It occurs randomly and from my numerous attempts to catch I suspect it may be related to the use of CPTs (colormaps) but could never find the cause. And I tried.

File: src/common_options.jl [hunk 1]
```diff
@@ -2014,34 +2015,33 @@ function line_decorated_with_string(str::AbstractString; dist=0)::String
...
+	SEP is the char separator used when ARG is a tuple or array of numbers
+"""
+arg2str(arg::Nothing,       sep = '/') = ""
+arg2str(arg::Real,          sep = '/') = @sprintf("%.12g", arg)
+arg2str(arg::Symbol,        sep = '/') = string(arg)
+arg2str(arg::Array{<:Real}, sep = '/') = rstrip(join([string(x, sep) for x in arg]), sep)
+arg2str(arg::Tuple,         sep = '/') = rstrip(join([string(x, sep) for x in arg]), sep)
+function arg2str(arg::Bool, sep = '/')
+    arg && return ""
+end
```
Comment (ZMAlt) [inline src/common_options.jl:2034]: What should we do if `arg` is `false`? Still throw an error?

**Comment by ZMAlt** [comment]
I reverted some type annotations, but for others, `arg2str` can ensure their type is `String` and restrict the `Any` propagation, I don't think they need to be reverted. If there are inputs of other types in the future, we add methods to `arg2str`. What do you think?

File: src/common_options.jl (hunk 1 above)
Comment (joa-quim) [reply src/common_options.jl:2034]: I guess we throw an error. The code never call is with a `false` but there may be instances where a user may indirectly trigger it.

File: src/common_options.jl [hunk 2]
```diff
@@ -247,7 +247,7 @@ function build_opt_R(val, symb::Symbol=Symbol())::String		# Generic function tha
 			_val::Vector{<:Float64} = vec(Float64.(collect(val)))
 			R = " -R" * @sprintf("%.15g/%.15g/%.15g/%.15g+r", _val[1], _val[3], _val[2], _val[4])::String
 		else
-			R = " -R" * rstrip(arg2str(val)::String, '/')		# Remove last '/'
```
Comment (joa-quim) [inline src/common_options.jl:250]: Why are doing this? That last slash needs to be removed.

File: src/common_options.jl [hunk 3]
```diff
@@ -1374,13 +1374,13 @@ parse_bo(d::Dict, cmd::String) = parse_b(d, cmd, [:b :bo :binary_out], "o")
...
 function parse_c(d::Dict, cmd::String)::Tuple{String, String}
 	# Most of the work here is because GMT counts from 0 but here we count from 1, so conversions needed
-	opt_val::String = ""
+	opt_val = ""
 	if ((val = find_in_dict(d, [:c :panel])[1]) !== nothing)
-		if (isa(val, Tuple) || isa(val, Array{<:Real}) || isa(val, Integer))
+		if isa(val, Tuple) || isa(val, Array{<:Real}) || isa(val, Integer)
 			opt_val = arg2str(val .- 1, ',')
 		elseif (isa(val, String) || isa(val, Symbol))
-			_val::String = string(val)		# In case it was a symbol
```
Comment (joa-quim) [inline src/common_options.jl:1382]: Don't remove these annotations. That's what insures ``_val`` or ``opt_val`` above will be `String` for sure.
``string(val)`` will be ``Any``, not ``String`` as one would think.

File: src/common_options.jl (hunk 2 above)
Comment (ZMAlt) [reply src/common_options.jl:250]:
```julia
arg2str(arg::Array{<:Real}, sep = '/') = rstrip(join([string(x, sep) for x in arg]), sep)
arg2str(arg::Tuple,         sep = '/') = rstrip(join([string(x, sep) for x in arg]), sep)
```
Done in arg2str

Comment (joa-quim) [reply src/common_options.jl:250]: OK.

**Comment by ZMAlt** [comment]
```julia
arg2str(arg::Array{<:Real}, sep = '/') = rstrip(join([string(x, sep) for x in arg]), sep)
arg2str(arg::Tuple,         sep = '/') = rstrip(join([string(x, sep) for x in arg]), sep)
```
`rstrip` will lead to `SubString{String} type`, I changed it to `String` in latest commit. After this, I did some checks.

1. `JET.jl`
- with "type annotations"

```julia
julia> @report_opt GMT.parse_c(Dict(:c => (1,2)), "")
No errors detected

julia> @report_opt GMT.parse_c(Dict(:c => [1,2]), "")
No errors detected

julia> @report_opt GMT.parse_c(Dict(:c => "1,2"), "")
═════ 1 possible error found ═════
┌ parse_c(d::Dict{Symbol, String}, cmd::String) @ GMT /Users/zm/Desktop/git/GMT.jl/src/common_options.jl:1384
│┌ parse(::Type{Int64}, s::String) @ Base ./parse.jl:253
││┌ parse(::Type{Int64}, s::String; base::Nothing) @ Base ./parse.jl:254
│││ runtime dispatch detected: convert(%1::Type{Int64}, %5::Union{Nothing, Int64})::Int64
││└────────────────────

julia> @report_opt GMT.parse_c(Dict(:c => 1), "")
No errors detected

julia> @report_opt GMT.parse_c(Dict(:c => "1"), "")
═════ 1 possible error found ═════
┌ parse_c(d::Dict{Symbol, String}, cmd::String) @ GMT /Users/zm/Desktop/git/GMT.jl/src/common_options.jl:1384
│┌ parse(::Type{Int64}, s::String) @ Base ./parse.jl:253
││┌ parse(::Type{Int64}, s::String; base::Nothing) @ Base ./parse.jl:254
│││ runtime dispatch detected: convert(%1::Type{Int64}, %5::Union{Nothing, Int64})::Int64
││└────────────────────

```

- without "type annotations"
```julia
julia> @report_opt GMT.parse_c(Dict(:c => (1,2)), "")
No errors detected

julia> @report_opt GMT.parse_c(Dict(:c => [1,2]), "")
No errors detected

julia> @report_opt GMT.parse_c(Dict(:c => "1,2"), "")
═════ 1 possible error found ═════
┌ parse_c(d::Dict{Symbol, String}, cmd::String) @ GMT /Users/zm/Desktop/git/GMT.jl/src/common_options.jl:1384
│┌ parse(::Type{Int64}, s::String) @ Base ./parse.jl:253
││┌ parse(::Type{Int64}, s::String; base::Nothing) @ Base ./parse.jl:254
│││ runtime dispatch detected: convert(%1::Type{Int64}, %5::Union{Nothing, Int64})::Int64
││└────────────────────

julia> @report_opt GMT.parse_c(Dict(:c => 1), "")
No errors detected

julia> @report_opt GMT.parse_c(Dict(:c => "1"), "")
═════ 1 possible error found ═════
┌ parse_c(d::Dict{Symbol, String}, cmd::String) @ GMT /Users/zm/Desktop/git/GMT.jl/src/common_options.jl:1384
│┌ parse(::Type{Int64}, s::String) @ Base ./parse.jl:253
││┌ parse(::Type{Int64}, s::String; base::Nothing) @ Base ./parse.jl:254
│││ runtime dispatch detected: convert(%1::Type{Int64}, %5::Union{Nothing, Int64})::Int64
││└────────────────────

```

There are no differences between with annotations and without annotations. As to `runtime dispatch detected: convert(%1::Type{Int64}, %5::Union{Nothing, Int64})::Int64`, I guess it comes from `parse` funciton whose inputs may be String or Nothing. Have nothing to do with type annotations.

2. `@code_warntype` gives similar results, but it's too long so I won't put it here.

File: src/common_options.jl (hunk 3 above)
Comment (ZMAlt) [reply src/common_options.jl:1382]: See my comments below, please.

File: src/common_options.jl (hunk 1 above)
Comment (ZMAlt) [reply src/common_options.jl]: Done

**Comment by joa-quim** [comment]
OK, I think we are converging to a solution where the `arg2str` is guarantied to return a string, but your JET test are not testing what really happen in the code. The ground problem is the `d` dictionary is a ``Dict{:Symbol, Any}`` so any values we extract out of it will be a `Any`. That is, if we remove the type annotations here
```
			if (length(d[key[k]]) == 2)		# Run the function
					cmd *= string(d[key[k]][1])::String * d[key[k]][2](Dict(key[k] => nt[k]), [key[k]])::String
```
the output of ``string(d[key[k]][1])`` is a ``Any`` not a ``String``, so those type annotations should be kept. When you tested with the contents of  ``Dict(:c => "1,2")``, sure that is a ``String`` and JET will be satisfied.

**Comment by ZMAlt** [comment]
> When you tested with the contents of `Dict(:c => "1,2")`, sure that is a `String` and JET will be satisfied.

Make sense.  I will redo a new PR and close this.

**Pull request closed without merging**